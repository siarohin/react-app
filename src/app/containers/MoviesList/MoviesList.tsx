import React, { ComponentType, ReactElement } from "react";
import { useHistory, useLocation, useParams } from "react-router-dom";
import { Helmet } from "react-helmet";
import { ActionCreatorWithoutPayload } from "@reduxjs/toolkit";
import { connect } from "react-redux";
import map from "lodash/map";
import isNil from "lodash/isNil";
import isEmpty from "lodash/isEmpty";

import { Grid } from "../../shared";
import {
  MoviesModels,
  State,
  UserPreferencesActions,
  RouterPath,
  UserPreferencesModels,
  SharedModels,
  MetaTitle,
  Endpoints
} from "../../core";
import { useComponentDidMount } from "../../hooks";
import { MovieCard } from "../../components";
import { getSearchValueFromPath } from "../../utils";
import NoMovies from "../../components/NoMovies";

const MoviesList = ({
  movies,
  searchValue,
  dispatch
}: {
  movies: Array<MoviesModels.IMovie>;
  searchValue: string;
  dispatch: (action: SharedModels.ActionWithPayload<any>) => ActionCreatorWithoutPayload;
}): ReactElement => {
  const history = useHistory();
  const { search } = useParams<{ search: string }>();
  const { pathname } = useLocation();

  // Fix for `Nested or latter components will override duplicate changes` with react-helmet
  const isMoviePath: boolean = pathname.match(/\/film/)?.[0] === RouterPath.Movie;
  const hasMoviesLength = movies?.length > 0;

  useComponentDidMount(() => {
    const isInitialLoading: boolean = isNil(searchValue) && isEmpty(movies);

    if (isInitialLoading) {
      const selected: string = getSearchValueFromPath(search);
      dispatch(UserPreferencesActions.updateSearchValue({ selected }));
    }
  });

  const handleEditableAction = (dialogAction: UserPreferencesModels.IMovieAction): void => {
    dispatch(UserPreferencesActions.updateDialogAction({ dialogAction }));
  };

  const handleMovieClick = (selectedMovie: MoviesModels.IMovie): void => {
    dispatch(UserPreferencesActions.updateSelectedMovieSuccess({ selectedMovie }));
    history.push(`${RouterPath.Movie}/${selectedMovie.id}`);
  };

  const MoviesContainer = ({ helmet }: { helmet: ReactElement<typeof Helmet> }): ReactElement => {
    return (
      <>
        {isMoviePath ? null : helmet}
        <Grid item container xs={12} spacing={2}>
          {map(movies, (movie) => (
            <Grid key={movie?.id} item xs={4}>
              <MovieCard onClick={handleMovieClick} editableAction={handleEditableAction} movie={movie} />
            </Grid>
          ))}
        </Grid>
      </>
    );
  };

  const NoMoviesContainer = ({ helmet }: { helmet: ReactElement<typeof Helmet> }): ReactElement => {
    return (
      <>
        {isMoviePath ? null : helmet}
        <NoMovies />
      </>
    );
  };

  return hasMoviesLength ? (
    <MoviesContainer
      helmet={
        <Helmet>
          <title>
            {search
              ? `${getSearchValueFromPath(search)?.toUpperCase()} - ${movies?.length} ${MetaTitle.SearchResult}`
              : MetaTitle.Search}
          </title>
          <link
            rel="canonical"
            href={`${Endpoints.ClientHost}${RouterPath.Search}/${encodeURI(getSearchValueFromPath(search))}`}
          />
        </Helmet>
      }
    />
  ) : (
    <NoMoviesContainer
      helmet={
        <Helmet>
          <title>{`${getSearchValueFromPath(search)?.toUpperCase()} - ${MetaTitle.NoMovies}`}</title>
          <link
            rel="canonical"
            href={`${Endpoints.ClientHost}${RouterPath.Search}/${encodeURI(getSearchValueFromPath(search))}`}
          />
        </Helmet>
      }
    />
  );
};

const mapStateToProps = (state: State): { movies: Array<MoviesModels.IMovie>; searchValue: string } => ({
  movies: state.movieList.movies,
  searchValue: state.userPreferences.search.selected
});

export default connect(mapStateToProps)(MoviesList as ComponentType<any>);
