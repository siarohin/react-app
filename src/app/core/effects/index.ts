import { combineEpics } from "redux-observable";

import { movieEffects } from "./movies.effect";
import { userPreferencesEffects } from "./user-preferences.effect";
import { toastEffects } from "./toast.effect";

export * from "./common.util";

export const effects = combineEpics(movieEffects, userPreferencesEffects, toastEffects);
