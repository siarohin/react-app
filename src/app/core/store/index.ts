export * from "./reducers";
export { SharedModels } from "./shared";
export { MoviesActions, MoviesModels } from "./movies";
export { UserPreferencesActions, UserPreferencesModels } from "./user-preferences";
export { ToastActions } from "./toast";
