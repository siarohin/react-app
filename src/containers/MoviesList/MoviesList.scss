*[class*="MuiCircularProgress-root"].app-movies-list__progress-bar {
  justify-self: center;

  *[class*="MuiCircularProgress-svg"] {
    color: var(--app-theme-data-viz-red);
  }
}
