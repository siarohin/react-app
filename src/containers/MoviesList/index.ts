import MoviesList from "./MoviesList";

export * from "./models";
export { MoviesList };
