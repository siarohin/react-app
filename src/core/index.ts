export * from "./constants";
export * from "./services";
export * from "./effects";
export * from "./store";
export * from "./store-configuration";
