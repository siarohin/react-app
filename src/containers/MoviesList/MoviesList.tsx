import React, { ComponentType, ReactElement } from "react";
import { connect } from "react-redux";
import map from "lodash/map";

import { Grid } from "../../shared";
import { MoviesModels, State, UserPreferencesActions, SharedModels } from "../../core";
import { MovieCard } from "../../components";
import { IMoviesListProps } from "./models";

const MoviesList = (
  props: Partial<MoviesModels.IMoviesState> & IMoviesListProps & SharedModels.IDispatchAction
): ReactElement<IMoviesListProps> => {
  const { movies, dispatch, editableAction } = props;

  const handleMovieClick = (selectedMovie: MoviesModels.IMovie): void => {
    dispatch(UserPreferencesActions.updateSelectedMovie({ selectedMovie }));
  };

  return (
    <Grid item container xs={12} spacing={2}>
      {map(movies, (movie) => (
        <Grid key={movie?.id} item xs={4}>
          <MovieCard onClick={handleMovieClick} editableAction={editableAction} movie={movie} />
        </Grid>
      ))}
    </Grid>
  );
};

const mapStateToProps = (state: State): Partial<MoviesModels.IMoviesState> => ({
  movies: state.movieList.movies
});

export default connect(mapStateToProps)(MoviesList as ComponentType<any>);
