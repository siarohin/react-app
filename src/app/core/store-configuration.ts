import { applyMiddleware, createStore, compose, Store } from "redux";
import { combineEpics, createEpicMiddleware, Epic } from "redux-observable";

import { reducers, State } from "./store";
import { isServer } from "./constants";

function getComposeEnhancers(): any {
  if (!isServer) {
    return (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
  }
  return compose;
}

export interface AppStore {
  store: Store;
  runEffects: (effects$?: Array<Epic>) => void;
}

/**
 * Store configuration
 */
export function configureStore(initialState: State = {} as State): AppStore {
  const epicMiddleware = createEpicMiddleware();
  const composeEnhancers = getComposeEnhancers();
  const middlewares = [epicMiddleware];

  const store: Store = createStore(reducers, initialState, composeEnhancers(applyMiddleware(...middlewares)));

  return {
    store,
    runEffects: (effects$?: Array<Epic>) =>
      effects$
        ? epicMiddleware.run(combineEpics(...effects$))
        : // eslint-disable-next-line no-console
          console.warn("Effects are not running, no epics array was provided")
  };
}
