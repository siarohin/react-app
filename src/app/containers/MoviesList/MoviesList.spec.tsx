import React from "react";
import { Provider } from "react-redux";
import { Route, BrowserRouter as Router } from "react-router-dom";
import renderer, { ReactTestRenderer } from "react-test-renderer";
import configureStore from "redux-mock-store";

import { State } from "../../core";
import MoviesList from "./MoviesList";

const mockStore = configureStore([]);

describe("Containers.MoviesList.MoviesList: ", () => {
  let store: any;
  let component: ReactTestRenderer;

  jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useParams: jest.fn().mockReturnValue({ search: "Search" })
  }));

  beforeEach(() => {
    store = mockStore({
      movieList: {
        movies: [{}]
      },
      userPreferences: {
        search: {
          selected: ""
        }
      }
    } as State);

    component = renderer.create(
      <Provider store={store}>
        <Router>
          <Route path="/search" component={MoviesList} />
        </Router>
      </Provider>
    );
  });

  it("should render component", () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
});
