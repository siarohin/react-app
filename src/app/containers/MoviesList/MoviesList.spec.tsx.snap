// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Containers.MoviesList.MoviesList:  should render component 1`] = `null`;
